src/components/input/DurationInput.js

```diff
@@ -0,0 +1,404 @@
+          <IconChevronUpSmallOutline />
+        </Icon>
+        <Icon className={theme['spinner-down']} {...spinnerDownProps} {...iconProps}>
+          <IconChevronDownSmallOutline />
+        </Icon>
+      </div>
+    );
+  }
+}
+
+class DurationInput extends PureComponent {
+  state = {
+    hours: 0,
+    minutes: 0,
+    seconds: 0,
+    hoursInputHasfocus: false,
+    minutesInputHasfocus: false,
+    secondsInputHasfocus: false,
+  };
+
+  updateStep = (type, n) => {
+    const { hours, minutes, seconds } = this.state;
+    let newHours = hours;
+    let newMinutes = minutes;
+    let newSeconds = seconds;
+
+    switch (type) {
+      case 'hours':
+        if (hours >= 0 && n >= 0) {
+          newHours = hours + n;
+        } else if (hours >= Math.abs(n) && n < 0) {
+          newHours = hours + n;
+        } else if (hours <= Math.abs(n) && n < 0) {
+          newHours = 0;
+        }
+        this.setState({ hours: newHours }, () => this.props.onChange(durationToSeconds(newHours, minutes, seconds)));
+        break;
+
+      case 'minutes':
+        if (minutes + n >= 60) {
+          newHours = hours + 1;
+          newMinutes = minutes - 60 + n;
+        } else if (n >= 0) {
+          newMinutes = minutes + n;
+          newHours = hours;
+        } else if (n < 0 && hours > 0 && minutes <= Math.abs(n)) {
+          newMinutes = minutes + 60 + n;
+          newHours = hours - 1;
+        } else if (n < 0 && minutes >= Math.abs(n)) {
+          newMinutes = minutes + n;
+        } else if (n < 0 && minutes <= Math.abs(n)) {
+          newMinutes = 0;
+        }
+        this.setState({ minutes: newMinutes, hours: newHours }, () =>
+          this.props.onChange(durationToSeconds(newHours, newMinutes, seconds)),
+        );
+        break;
+
+      case 'seconds':
+        if (n > 0 && minutes * 60 + seconds >= 3600 - Math.abs(n)) {
+          newSeconds = seconds - 60 + n;
+          newMinutes = Math.floor(newSeconds / 60);
+          newHours = hours + 1;
+        } else if (seconds + n >= 60) {
+          newMinutes = minutes + 1;
+          newSeconds = seconds - 60 + n;
+        } else if (n >= 0) {
+          newSeconds = seconds + n;
+          newMinutes = minutes;
+        } else if (n < 0 && hours >= 1 && minutes === 0 && seconds + n < 0) {
+          newSeconds = 60 + n;
+          newMinutes = 59;
+          newHours = hours - 1;
+        } else if (n < 0 && minutes > 0 && seconds < Math.abs(n)) {
+          newSeconds = seconds + 60 + n;
+          newMinutes = minutes - 1;
+        } else if (n < 0 && seconds >= Math.abs(n)) {
+          newSeconds = seconds + n;
+        } else if (n < 0 && seconds <= Math.abs(n)) {
+          newSeconds = 0;
+        }
+        this.setState({ hours: newHours, seconds: newSeconds, minutes: newMinutes }, () =>
+          this.props.onChange(durationToSeconds(hours, newMinutes, newSeconds)),
+        );
+        break;
+
+      default:
+        break;
+    }
```

Comment:
Something that might help simplify these calculations could be to hold a single value in your state, this would be the total amount of seconds which represents the duration. In the `updateStep` function we should then be able to add the `n` to what's in the state. (`n * 60 * 60` for type hours, `n * 60` for minutes and `n` for seconds).

Then we can move the work of splitting this into hours/minutes/seconds to our render function to show them in the separate input fields. An easy trick to do this is to use rest calculations with the modulo operator (`%`).

Something like this could work:
```
const secondsInMinute = 60;
const secondsInHour = secondsInMinute * 60;

const splitSecondsIntoHoursMinutesSeconds = seconds => {
	const hours = Math.floor(seconds / secondsInHour);
	const minutes = Math.floor((seconds % secondsInHour) / secondsInMinute);
	const remainingSeconds = seconds % secondsInMinute;
	return {
		hours,
		minutes,
		seconds: remainingSeconds,
	};
};
```

Reply:
Damn, I was writing this down 

So. I agree 😃

Reply:
Made some comments on it as well, but this indeed seems to be the best solution 👍